File: server/src/main/java/com/nexoffer/controller/ResumeQuestionsController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/resume-questions")
public class ResumeQuestionsController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public ResumeQuestionsController(ProfileRepository profileRepository,
                                    HistoryRepository historyRepository,
                                    LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/generate")
    public ResponseEntity<?> generateResumeQuestions(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String resumeText = body != null ? body.get("resumeText") : null;
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (resumeText == null || resumeText.isBlank()) resumeText = p.getResumeText();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
        }

        if (resumeText == null || resumeText.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Resume text is missing. Please upload your resume in Profile first."));
        }

        String targetCompany = (company != null && !company.isBlank()) ? company : "Target Company";
        String targetRole = (role != null && !role.isBlank()) ? role : "Software Developer";

        String prompt = "Carefully inspect the candidate's Resume provided in context.\n" +
                "Generate high-impact technical interview questions specifically customized to their resume content in 3 areas:\n" +
                "1. projectQuestions (3-4 questions targeting architecture, trade-offs, security, database)\n" +
                "2. skillQuestions (2-3 deep questions probing language fundamentals and tools)\n" +
                "3. experienceQuestions (2 questions on team collaboration, agile, code review)\n\n" +
                "Return ONLY valid JSON with this schema:\n" +
                "{\n" +
                "  \"projectQuestions\": [{ \"question\": \"...\", \"category\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"skillQuestions\": [{ \"question\": \"...\", \"category\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"experienceQuestions\": [{ \"question\": \"...\", \"category\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"suggestedAnswer\": \"...\" }]\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("resumeText", resumeText);
        context.put("company", targetCompany);
        context.put("role", targetRole);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are an expert technical interviewer who rigorously analyzes candidate resumes to ask pointed, architecture-level questions.",
                context,
                true,
                "resume-questions"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("resume-questions", context);

        historyRepository.save(new History(
                user.getId(),
                "Resume Questions",
                targetCompany,
                targetRole,
                "Generated resume-specific project & skill questions"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}

File: server/controllers/resumeQuestionsController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Generate deep-dive questions based on the candidate's resume
// @route   POST /api/resume-questions/generate
const generateResumeQuestions = async (req, res, next) => {
  try {
    let { resumeText, company, role } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!resumeText) resumeText = profile.resumeText;
      if (!company) company = profile.company;
      if (!role) role = profile.role;
    }

    if (!resumeText || resumeText.trim().length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Resume text is missing. Please upload your resume in Profile first.',
      });
    }

    const prompt = `
Carefully inspect the candidate's Resume provided in context.
Extract specific details about their projects, tech stack choices, architecture, experience, and listed skills.
Generate high-impact technical interview questions specifically customized to their resume content in 3 areas:
1. "projectQuestions" (3-4 questions targeting project architecture, trade-offs, security, database, and hardest challenges)
2. "skillQuestions" (2-3 deep questions probing language fundamentals and tools listed on the resume)
3. "experienceQuestions" (2 questions on team collaboration, agile methodology, or work experience)

Return ONLY valid JSON with this schema:
{
  "projectQuestions": [
    { "question": "...", "category": "...", "difficulty": "Easy|Medium|Hard", "suggestedAnswer": "..." }
  ],
  "skillQuestions": [
    { "question": "...", "category": "...", "difficulty": "Easy|Medium|Hard", "suggestedAnswer": "..." }
  ],
  "experienceQuestions": [
    { "question": "...", "category": "...", "difficulty": "Easy|Medium|Hard", "suggestedAnswer": "..." }
  ]
}
`;

    const context = {
      resumeText,
      company: company || 'Target Company',
      role: role || 'Software Developer',
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'resume-questions',
      systemPrompt: 'You are an expert technical interviewer who rigorously analyzes candidate resumes to ask pointed, architecture-level questions.',
    });

    const data = aiResult.parsedJson || llmGateway.generateFallbackResponse('resume-questions', context);

    await HistoryRepository.create({
      userId: req.user._id,
      module: 'Resume Questions',
      company: company || 'General',
      role: role || 'General',
      summary: 'Generated resume-specific project & skill questions',
      metadata: { projectCount: data.projectQuestions?.length || 0 },
    });

    res.status(200).json({
      success: true,
      data,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  generateResumeQuestions,
};
